Editor: Display Site Icon (if one is set) in Gutenberg Fullscreen Mode

jeyip opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
This PR addresses issue #20929  `Iterations on W icon in the editor header`. It renders the current site icon (if one is set) instead of the default WordPress icon in the site **and** post editor.

**Note**:
1. @noahtallen made a great callout. This PR only handles displaying a user uploaded site icon. We may need to update the site icon from the Gutenberg editor in the future as well. Although not addressed here, a strategy for updating site icons is worthwhile to consider while reviewing this PR.
2. Since information about site icons wasn't originally provided to the Gutenberg editor, we exposed the site icon url as a settings field. We use `register_setting` to extend the WordPress `wp/v2/settings` API. @ockham discussed the idea of including that as a field in Core WordPress itself. We are considering opening a new issue, and I'd love to hear more feedback about this.

## How has this been tested?
- Set up local instance of WordPress and Gutenberg by following instructions [here](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/getting-started.md)
- Navigate to WordPress admin sidebar -> _Appearance_ -> _Customize_ -> _Site Identity_
- Upload a site icon
- Publish changes
- If already enabled, navigate to the site editor
  - Verify changes in full-screen mode
- If not already enabled, navigate to WordPress admin sidebar -> _Gutenberg_ -> _Experiments_
  - Enable `Full Site Editing`
  - Enable `Full Site Editing Demo Templates`
  - Navigate to the site editor and verify changes in full-screen mode


<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots
### With no uploaded site icon
<img width="1536" alt="Screen Shot 2020-06-05 at 1 07 29 PM" src="https://user-images.githubusercontent.com/5414230/83925247-8be26300-a73b-11ea-847c-de10ce479766.png">

### With site icon
<img width="1404" alt="Screen Shot 2020-06-08 at 6 24 13 PM" src="https://user-images.githubusercontent.com/5414230/84096006-a293ee80-a9b5-11ea-9110-82fefad81490.png">

### GIF
![Site Icon](https://user-images.githubusercontent.com/5414230/84096135-f999c380-a9b5-11ea-85da-2c288d68df9a.gif)

## Types of changes
- Expose site icon URL to Gutenberg editor
- Display site icon in full-screen mode

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->
### Requirements
**With default WordPress icon**
- [ ] Icon button disappears on screens smaller than `782px`
- [ ] Clicking icon button redirects user to wp-admin home page
- [ ] Back button tooltip displays on hover after timed delay

**With custom site icon**
- [ ] Icon button disappears on screens smaller than `782px`
- [ ] Clicking icon button redirects user to wp-admin home page
- [ ] Back button tooltip displays on hover after timed delay
- [ ] Icon is displayed for Contributor, Author, Editor, and Admin roles.

### Browsers
- [ ] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari
- [ ] IE11

Commit:
Register setting for site icon url

Info about site icon was previously unavailable in the Gutenberg editor. Registering a setting addresses the problem by adding site icon url as a field to the wp/v2/settings REST API. The information can then be fetched by the client.

noahtallen commented on gutenberg.php:
```diff
@@ -161,3 +161,16 @@ function gutenberg_rest_nonce() {
+ * Exposes the site icon url to the Gutenberg editor through the WordPress REST API.
+ */
+function register_site_icon_url_setting() {
+	register_setting('general', 'site_icon_url', array(
```
we should update this to match the WordPress php style standards. I think you can also find issues locally by running the lint php script. (`npm run lint-php`) Basically you just need to add spaces before and after the function call

ockham replied:
PHPCBF is our friend:

```
composer install
vendor/bin/phpcbf gutenberg.php
```

ockham replied:
(There are integrations for popular editors. I like to flip on format-on-save :grimacing: )

jeyip committed:
Display user uploaded site icon for edit site if available

jeyip committed:
Apply default site icon styling only to default icon

noahtallen reviewed (commented):
This seems like a good approach to me for now. As we discussed earlier, the main concern is that updating the `site_icon_url` setting would not actually change the `site_icon` option under the hood. All of the site icon related code (like the favicon and `get_site_icon_url` pulls from that option under the hood. Unfortunately, since that option stores a media ID and not a URL, we can't use it directly for our purposes here.

ockham commented on packages/edit-site/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -2,11 +2,22 @@
+	const [ siteIconURL ] = useEntityProp( 'root', 'site', 'site_icon_url' );
+	
+	const isRequestingSiteIcon = useSelect( ( select ) => {
+		return ! select( 'core/data' ).hasFinishedResolution(
```
I normally use `isResolving` which I _think_ is just `! hasFinishedResolution` (haven't checked tho).

```suggestion
		return select( 'core/data' ).isResolving(
```

ockham commented on packages/edit-site/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -2,11 +2,22 @@
+		return ! select( 'core/data' ).hasFinishedResolution(
+			'core',
+			'getEntityRecord',
+			[ 'root', 'site', undefined ]
```
Doesn't this need to match the [`useEntityProp` args](https://github.com/WordPress/gutenberg/pull/22952/files#diff-5bb3dbd3c539d837e40778fd5c4e55baR11)?

jeyip replied:
Great question. I'm not entirely sure because I couldn't find great documentation on `hasFinishedResolution`, but I'll investigate and update you here.

noahtallen replied:
one thought is that since `'root', 'site'` translates to a single query which fetches `/settings`, there probably isn't a need to be more granular, since every prop is resolved in the same request. (I think)

ockham commented on packages/edit-site/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -15,7 +26,8 @@ function FullscreenModeClose( { icon } ) {
 
-	const buttonIcon = icon || wordpress;
+	const shouldDisplaySiteIcon = siteIconURL || isRequestingSiteIcon;
+	const buttonIcon = shouldDisplaySiteIcon ? null : icon || wordpress;
```
TBH I don't know the order of precedence for a ternary vs a `||`. Can we rewrite this a bit to make it clearer?

jeyip replied:
I totally agree 👍 . This is something I wanted to clean up (along with styling) before a final code review.

jeyip replied:
I clean up the conditional logic [here](https://github.com/WordPress/gutenberg/pull/22952/commits/1b23ed352d612ad0d235d973c0c1b5fd27bd17d5). Let me know if that's easier to read.

paaljoachim commented:
I tried testing on gutenberg.run but was not able to because of errors in the customizer.
It is awesome that you are working on this! I will see if I can find another way to check out the PR.

jeyip committed:
Remove unnecessary whitespace

jeyip committed:
Fix PHP formatting with linter

jeyip committed:
Use isResolving to determine site icon loading state

Previously, we were using hasFinishedResolution and inverting the resulting boolean to determine site icon loading state.

Implementing isResolving instead let's us determine the same loading state without boolean inversion.

noahtallen commented on packages/edit-site/src/components/header/fullscreen-mode-close/style.scss:
```diff
@@ -8,22 +8,23 @@
 		align-items: center;
 		align-self: stretch;
 		border: none;
-		background: #23282e; // WP-admin gray.
```
per the design, should we keep this as the background even when we're using the site icon?

jeyip replied:
That's a great point 👍. I'll fix it and maintain the WP-admin gray even with the site icon.

jeyip committed:
Add unit tests for edit-site

jeyip committed:
Refactor button icon logic

Previously, to render the default icon, we were passing an SVG through the icon prop to the @wordpress/components Button. The data provided for user uploaded site icons, however, are incompatible. We receive custom URLs, and not SVGs, for custom site icons.

Rather than support the logic for both Button icon props and site icon image tags, we do away with the Button icon prop entirely.

We now explicitly define the children for Button, which consolidates the display logic for both default and custom site icons.

jeyip committed:
Update styling for site icon

jeyip committed:
Add user uploaded site icon to post editor

youknowriad commented on packages/edit-post/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -7,12 +7,23 @@ import { get } from 'lodash';
 import { wordpress } from '@wordpress/icons';
 
 function FullscreenModeClose() {
+	const [ siteIconURL ] = useEntityProp( 'root', 'site', 'site_icon_url' );
```
I think useEntityProp is meant to be used in conjunction with a context provider around the block editor (so more for blocks) or is this a correct usage @epiqueras ?

Why not just `wp.data.select('core').getSite()` in the useSelect

youknowriad replied:
also, should we add this API request to the preloaded paths?

epiqueras replied:
Yeah, we should use `getEntityRecords( 'root', 'site' ).site_icon_url` to make it simpler and decoupled from any providers.

noahtallen replied:
interesting :) Is this documented behavior? It's a confusing difference IMO

jeyip replied:
> should we add this API request to the preloaded paths?

@youknowriad Would you be able to elaborate more on what it means to add this to preloaded paths?

youknowriad replied:
yes, basically if you add a path to this array, the server will inject  the response on the initial page load avoiding to trigger a request that we  know will always be triggered on that page.

A similar array exists for edit-post too.

jeyip replied:
[As suggested by Timothy](https://github.com/WordPress/gutenberg/pull/22952#discussion_r438124552), we now use the WordPress REST index to register site icon url. It looks like the index is already being preloaded [here](https://github.com/WordPress/gutenberg/blob/master/lib/edit-site-page.php#L190), but let me know if there's more to do.

youknowriad commented on gutenberg.php:
```diff
@@ -161,3 +161,20 @@ function gutenberg_rest_nonce() {
+	);
+}
+
+add_action( 'rest_api_init', 'register_site_icon_url_setting' );
```
Let's make sure we follow-up with a trac ticket here. cc @TimothyBJacobs

mtias replied:
Also what is the permissions situation with this setting? Since it's only for display we should ensure it's accessible to all roles.

jeyip replied:
Thanks for calling out permissions @mtias . All roles that can access the editor but that cannot edit site configuration (Contributor, Author, and Editor) receive a 403 response when querying site information. No user uploaded site icon is displayed except for Admins and Super Admins.

Correct me if I'm wrong, but it seems like exposing site icon Url through the WordPress REST API isn't possible unless we alter the permissions of the `wp/v2/settings` endpoint. This seems like a non-starter. 

@noahtallen and I had discussed other approaches. For example, instead of using the REST API, we might inject site icon Url as a prop on plugin initialization. Let me know if you have other ideas.

jeyip replied:
I implemented the `rest_index` filter. The site icon is now viewable by all roles (except subscriber, which can't access the editor)

youknowriad commented on packages/edit-post/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -28,16 +39,30 @@ function FullscreenModeClose() {
 				post_type: postType.slug,
 			} ) }
 			label={ get( postType, [ 'labels', 'view_items' ], __( 'Back' ) ) }
-		/>
```
If  I remember properly since now, this is not an icon button, we might want to add a `showTooltip` prop to show the tooltip.

jeyip replied:
Adding the `showTooltip` worked  👍

youknowriad reviewed (commented):
Nice work here. Left some comments.

jeyip committed:
Display tooltip for editor back button

jeyip committed:
Use getEntityRecord to fetch site icon URL for edit-site

useEntityProp was pulling information about site icon url from the core-data store. Although the method worked, it relied on a context provider around the block editor.

The context is unnecessary in this instance, and so, to prevent unnecessary coupling, we replace useEntityProp with getEntityRecord instead.

jeyip committed:
Use getEntityRecord to fetch site icon URL for edit-post

jeyip committed:
Register site icon url with rest index filter

All roles that could access the editor but that could not edit site configurations (Contributor, Author, and Editor) received a 403 response when querying the settings WordPress REST API. This is because site configurations and the settings API are only accessible to Admins and Super Admins.

We want the read-only site icon url to be accessible to all roles. In order to accomplish that, we opt to register the site icon url with the WordPress REST index endpoint instead of the WordPress REST settings endpoint.

jeyip committed:
Fetch site icon url from the WordPress REST index

jeyip committed:
Add inline documentation when registering site icon url

TimothyBJacobs reviewed (approved):
Looks good REST AP wise.

Addison-Stavlo commented:
Testing this on my end works well with and without the icon set.  This is great to see!

noahtallen commented on packages/edit-post/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -7,37 +7,63 @@ import { get } from 'lodash';
-			postType: getPostType( getCurrentPostType() ),
-		};
-	}, [] );
+			return {
```
@mtias @youknowriad do you think we should limit this to just edit-site for now, or do you think it would also be good to start testing it in edit-post?

mtias replied:
We should do it in both, but if we are not certain yet, better to keep it in edit-site. We would need to decide if it's ready for 5.5 as well (in edit-post).

jeyip replied:
> We would need to decide if it's ready for 5.5 as well

Sounds good 👍 . I'm newer on the team so I'm not sure what this entails. Could you elaborate?

Also, I wanted to call out that I'm on the gardening rotation so development on this PR is going to slow down until the end of the month.

mtias replied:
The second week of July is the beta period for WordPress 5.5. Anything that affects the post-editor needs to be in good shape by then.

jeyip replied:
@mtias I'm back from the gardening rotation, which leaves us 2 weeks to test this feature. 

Would you be able to send me documentation about how I'd make sure the post-editor is "in good shape" with these changes? Or is this something I'd coordinate with @sirreal?

jeyip replied:
@mtias @vindl Since it sounds like there is more extensive testing needed before releasing this feature in `edit-post`, I'm leaning towards separating this feature into two PRs. The current PR would be for `edit-site`, and the follow-up would be for `edit-post`. 

Since the work here isn't urgent, we wouldn't have to rush testing while we prep for 5.6. Let me know what you think.

paaljoachim commented:
I see some mentions above if we should add this to 5.5. It would be good with a pros and cons list.
But all in all I do think we could wait. As it is not an important feature as I understand it but one that is very nice to have. It could instead be added to the Gutenberg plugin after 7th of July. Giving space for a good testing period. Then added to 5.6.

youknowriad commented on packages/core-data/src/entities.js:
```diff
@@ -17,6 +17,12 @@ import { apiFetch, select } from './controls';
+		name: 'base',
+		kind: 'root',
+		baseURL: '',
+	},
```
The only think I'm unsure of in this PR is whether we should consider this endpoint as an entity or not? Are we already using this endpoint in Gutenberg? How are we doing so?

The alternative might be a dedicated action/selector on the 'core-data' store. Any thoughts @epiqueras

epiqueras replied:
How is this different from the "Site" entity?

youknowriad replied:
This is the "index" endpoint. These are alo site settings somehow but the endpoint is separate because this one is I believe authorized for more folks.

epiqueras replied:
Maybe we should call it Public Site?

epiqueras replied:
I think it makes sense as an entity if it has multiple properties we want to edit and save in unison.

TimothyBJacobs replied:
Yeah the `index` is completely public.

jeyip replied:
@epiqueras The index returns, among other properties, information including site name, site description, site icon url, and timezone string. These seem like properties we'd want to edit in the future. 

I'm uncertain, however, if that's possible through the current index endpoint.

epiqueras replied:
Ideally, we should have a single read/write endpoint for these "site" settings.

epiqueras replied:
Aren't all those properties already present in the settings endpoint?

jeyip replied:
I completely agree. We initially attempted to consolidate site icon url alongside the WP "site" settings REST endpoint, rather than through `index`. That endpoint, however, was only accessible to admin and super-admin roles. We want to display the site icon to all roles.

There's been [work](https://github.com/WordPress/gutenberg/pull/19967) and [discussion](https://wordpress.slack.com/archives/C02RQC26G/p1579198597407600) to provide read-only access to some site configuration properties, but [Timothy has said this won't be tackled until WordPress 5.6](https://github.com/WordPress/gutenberg/pull/22952#discussion_r438124552). I'm happy to try other ways of consolidating information if you have suggestions.

epiqueras replied:
Ahh, ok. I guess, for now, we can add this new entity, but let's prefix it with `__unstable` so that we can eventually consolidate things.

TimothyBJacobs replied:
> Ideally, we should have a single read/write endpoint for these "site" settings.

Hoping to get to this in 5.6

jeyip replied:
To clarify, when you say prefix with `__unstable`, do you mean the name like so?

```javascript
	{
		label: __( 'Base' ),
		name: '__unstableBase',
		kind: 'root',
		baseURL: '',
	},

```

epiqueras replied:
Yes

youknowriad reviewed (commented):
I don't have an icon and upon loading the page, I noticed a delay where the icon stays black entirely. is this because preloading is missing?

jeyip reviewed (commented):
> I don't have an icon and upon loading the page, I noticed a delay where the icon stays black entirely. is this because preloading is missing?

@youknowriad I had assumed the site icon was being preloaded because of our previous discussion [here](https://github.com/WordPress/gutenberg/pull/22952#discussion_r438479913) and the path preloading noted [here](https://github.com/WordPress/gutenberg/blob/master/lib/edit-site-page.php#L185-L186), but what you're seeing with the delay is what I've seen as well. If preloading means that there should be no delay at all, I'll need to investigate further.

**Update:** I can't seem to figure out why the provided paths aren't working. It seems like none of the paths are actually being preloaded. For example, [it looks like the taxonomies endpoint should be preloaded](https://github.com/WordPress/gutenberg/blob/master/lib/edit-site-page.php#L188). But when I navigate to the site editor, there's still a request made to the taxonomies endpoint. Any ideas @youknowriad?

<img width="1037" alt="Screen Shot 2020-06-29 at 9 43 32 PM" src="https://user-images.githubusercontent.com/5414230/86084562-0dcf5e80-ba52-11ea-995d-a9dff434cef5.png">

jeyip committed:
Add __unstable prefix to base entity

Because of limitations with permissioning, we are adding site icon url to the WP Rest index endpoint instead of the settings endpoint.

Once the settings endpoint is updated and accessible by more than the admin and super admin roles, we will remove the base entity (hence the unstable prefix). At that time, we will subsequently add site icon url to the settings endpoint.

youknowriad commented:
@jeyip After some tests, it seems the path that needs to be preloaded is `/?context=edit` which is not currently being preloaded. We do preload `/` instead. I guess we should preload the one with the "context" as it's how entities work but I'm also wondering if there's another place we trigger an API for `/` without context where we should unify with the entities call to avoid two requests for the same endpoint.

jeyip commented:
@youknowriad 🎉 Thanks for the help! That makes complete sense.

Would it be safe to say that, if `/` was being consumed elsewhere, we would be able to see an independent network request? After experimenting with my local instance, I see that whenever `/?context=edit` is excluded, Chrome records an extra request. When included, there is no additional request made.

When I exclude `/`, I see no additional requests to the REST API. I also smoke tested basic core FSE functionality and don't see a demonstrable difference.

jeyip committed:
Preload index endpoint

youknowriad commented:
>  Do you happen to know where the entry point for edit-post is?

It's currently in Core which means we need to add a new list on the plugin and add the "Backport to Core" label to this PR so we don't forget to include this Change in Core when we backport this Gutenberg version to core.

ockham commented on packages/edit-post/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -7,37 +7,64 @@ import { get } from 'lodash';
+	if ( siteIconUrl ) {
+		buttonIcon = (
+			<img
+				alt="site-icon"
```
Bit of a nitpick, but shouldn't the [`alt` text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) be more human- (rather than machine-) readable? (This might imply we need to translate it.)

E.g.

```suggestion
				alt={ __( 'Site Icon' ) }
```

ockham commented on packages/edit-post/src/components/header/fullscreen-mode-close/test/index.js:
```diff
@@ -0,0 +1,68 @@
+			} );
+
+			const { getByAltText } = render( <FullscreenModeClose /> );
+			const siteIcon = getByAltText( 'site-icon' );
```
Maybe a different kind of selector (e.g. class name?) could be more stable? ([See](https://github.com/WordPress/gutenberg/pull/22952/files#r448381310))

jeyip replied:
This makes sense since alt text may be dynamic now. 👍

ockham commented on packages/edit-post/src/components/header/fullscreen-mode-close/test/index.js:
```diff
@@ -0,0 +1,68 @@
+			const { container, queryByAltText } = render(
+				<FullscreenModeClose />
+			);
+			const siteIcon = queryByAltText( 'site-icon' );
```
Same as https://github.com/WordPress/gutenberg/pull/22952/files#r448381902?

ockham commented:
BTW in the long run, should we maybe try to converge the `edit-site` and `edit-post` versions of `FullscreenModeClose` into one shared component?

ockham commented on packages/edit-site/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -2,29 +2,61 @@
+	if ( siteIconUrl ) {
+		buttonIcon = (
+			<img
+				alt="site-icon"
```
See https://github.com/WordPress/gutenberg/pull/22952/files#r448381310

ockham commented on packages/edit-site/src/components/header/fullscreen-mode-close/test/index.js:
```diff
@@ -0,0 +1,64 @@
+			} );
+
+			const { getByAltText } = render( <FullscreenModeClose /> );
+			const siteIcon = getByAltText( 'site-icon' );
```
[see](https://github.com/WordPress/gutenberg/pull/22952/files#r448381902)

ockham commented on packages/edit-site/src/components/header/fullscreen-mode-close/test/index.js:
```diff
@@ -0,0 +1,64 @@
+			const { container, queryByAltText } = render(
+				<FullscreenModeClose />
+			);
+			const siteIcon = queryByAltText( 'site-icon' );
```
[see](https://github.com/WordPress/gutenberg/pull/22952/files#r448381902)

ockham reviewed (commented):
This is looking pretty good to me! I left a few minor notes.

I'm not giving final approval, since on my Linux box, I still cannot upload files to my local Gutenberg env, so I couldn't actually test the new feature :grimacing:

jeyip committed:
Translate alt text for site icon in edit post

jeyip committed:
Translate alt text for site icon in edit site

ockham commented:
> > Do you happen to know where the entry point for edit-post is?
> 
> It's currently in Core which means we need to add a new list on the plugin and add the "Backport to Core" label to this PR so we don't forget to include this Change in Core when we backport this Gutenberg version to core.

I think it's this: https://github.com/WordPress/WordPress/blob/f665e823a53f6b4205f37ff4e3222aacac3ed456/wp-admin/edit-form-blocks.php#L41-L53

which means we should be able to extend it in Gutenberg through the `block_editor_preload_paths` filter: https://github.com/WordPress/WordPress/blob/f665e823a53f6b4205f37ff4e3222aacac3ed456/wp-admin/edit-form-blocks.php#L55-L65.

IIUC, we need this if we want to continue to include the Site Icon for the Post Editor.

AFAICS, there are two ways forward:

1. Include that change, and keep the changes to `packages/edit-post/` in this PR.
2. Don't include that change, and drop the changes to `packages/edit-site` from this PR (or rather, move them to a follow-up PR).

I think that the second option might de-risk and accelerate this PR.

youknowriad commented:
I'd rather do 1 personally to keep consistency between the two screens.

jeyip commented:
@ockham Thanks for elaborating! That's super helpful.

@youknowriad If we'd prefer to keep changes to `edit-post` and `edit-site` in the same PR, is the approach Bernie is describing what you were describing earlier? If not, what does "It's currently in Core which means we need to add a new list on the plugin" mean?

youknowriad commented:
> is the approach Bernie is pointing out what you were describing earlier?

Yes, this should work 👍 Thanks.

jeyip committed:
Preload edit post index endpoint

Data from WordPress REST endpoints can be preloaded. We're effectively asking the server to inject a response on the initial page load that we know will always be triggered on that page. The alternative is waiting for the page to load, waiting for the scripts to executre, and then waiting for a response to a network request.

Preloading the site icon means that upon initial page load, rather than a blank placeholder, the icon will be displayed almost immediately. Edit post preloading happens in core, and the endpoints we specify for preloading are extensible through the block_editor_preload_paths filter.

This change preloads the endpoint that serves site icon data to edit post.

jeyip commented:
@ockham @youknowriad I modified the `block_editor_preload_paths` and we're now preloading the index endpoint for `edit-post`. Let me know if there's a better place to extend the `block_editor_preload_paths`

Thanks again for all of the feedback. It's much appreciated ⭐

youknowriad commented:
I think I'd add that to the "compat.php" to clarify that it's something that need to be included in Core list.

ockham commented:
> @ockham @youknowriad I modified the `block_editor_preload_paths` and we're now preloading the index endpoint for `edit-post`.

Awesome, thanks!

> * I thought about including the changes to `gutenberg.php`, but the purpose of preloading solely for `edit-post` seem distinct from the responsibilities of that file.
> * I ended up creating a separate php file to extend preload paths
> 
> Let me know if there's a better place or clearer way to change this. Thanks again for all of the feedback. It's much appreciated

Good question :thinking: One potential contender could be `lib/compat.php`: https://github.com/WordPress/gutenberg/blob/5078c43a3d72eded68d94488dbe6093454815a0e/lib/compat.php#L3-L6

So yeah, maybe it'd make sense to maybe move that filter into that file, so we don't have to introduce a new file :thinking:

jeyip committed:
Add core trac ticket annotation

jeyip commented:
I moved the edit post preloading into the `compat.php` file. @ockham would you mind adding the "Backport to Core" label to this ticket? (I'm not a contributor to Gutenberg yet)

noahtallen commented:
Done. Did we decide that we wanted this to land in the wp core release?

jeyip commented:
I'll defer to Riad. I'm happy to set this up for 5.5 or 5.6.

youknowriad commented:
yes, unfortunately, we're definitely passed the deadline for 5.5.

noahtallen commented:
that makes sense to me. 👍

Copons commented on packages/edit-post/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -7,37 +7,64 @@ import { get } from 'lodash';
+			/>
+		);
+	} else if ( isRequestingSiteIcon ) {
+		buttonIcon = null;
```
I'm wondering if it would make sense to display the WP logo while the site icon is loading, instead of showing nothing. 🤔 

Something like:
```jsx
const buttonIcon = siteIconUrl
	? <img src={ siteIconUrl } />
	: <Icon icon={ wordpress } />
```

I guess we wouldn't even need `isRequestingSiteIcon`, as `siteIconUrl` should be falsy while loading.

ockham replied:
I'm on the fence about this one. A similar thought had occurred to me, and it would certainly simplify the code, but then again: 

- UX: The WP logo might actually confuse people ("Hey, I set my own logo, why is it still showing the W thing?")
- IIUC, we'd be showing the placeholder while `siteIconUrl` is loading (rather than while the actual image file is being loaded from that URL).
- We're actually preloading the endpoint that gives us `siteIconUrl`, so the URL should be available near-instantly.

My vote for keeping as-is; this is something we can iterate on later.

Copons replied:
Makes sense! I'll resolve the conversation then 🙂

Copons commented on packages/edit-post/src/components/header/fullscreen-mode-close/index.js:
```diff
@@ -7,37 +7,64 @@ import { get } from 'lodash';
-			className="edit-post-fullscreen-mode-close"
-			icon={ wordpress }
-			iconSize={ 36 }
+			className="edit-post-fullscreen-mode-close has-icon"
```
I might be missing some context, but is there a reason why we're adding the `.has-icon` class?

ockham replied:
Good question. 

Previously, we used to pass the `wordpress` icon (an [`SVG` component](https://github.com/WordPress/gutenberg/blob/41cf77e0fbc52d4b57c1a477d8754dafbb0f4ddf/packages/icons/src/library/wordpress.js#L7-L9) from [`@wordpress/icons`](https://github.com/WordPress/gutenberg/pull/22952/files/c0b49d2a273eeeb4b2c7d731ce150bb171ef9c66#diff-5e54b8ed48d99cb46a591a186b9ad138L13)) via the [`icon` prop](https://github.com/WordPress/gutenberg/pull/22952/files/c0b49d2a273eeeb4b2c7d731ce150bb171ef9c66#diff-5e54b8ed48d99cb46a591a186b9ad138L34) to `Button`. The `Button` component forwards that prop to an [`Icon` element it embeds](https://github.com/WordPress/gutenberg/blob/41cf77e0fbc52d4b57c1a477d8754dafbb0f4ddf/packages/components/src/button/index.js#L113), and sets the [`has-icon` prop](https://github.com/WordPress/gutenberg/blob/41cf77e0fbc52d4b57c1a477d8754dafbb0f4ddf/packages/components/src/button/index.js#L63).

Since we're no longer passing the logo to `Button` via the `icon` prop (but wrap the Button around the logo), we're losing the `.has-icon` styling and have to add it back manually.

However, we might explore an alternative...

ockham replied:
The `Icon` component [seems to accept](https://github.com/WordPress/gutenberg/blob/bdfdd73f4bd9295057b9bfa51fc996185dedb44f/packages/components/src/icon/index.js#L62-L67) pretty much any valid component as a fallback. So we might be able to continue to pass in our custom log (or fallback) via `Button`'s `icon` prop, and remove the manually added `has-icon` prop (which is arguably an implementation thing rather than part of the API).

ockham replied:
I was experimenting a bit with this, and it would be possible. However, the `Icon` component is also passing a `size` prop to the element it clones, and `<img />` elements don't support that. So let's keep things as they are.

ockham commented:
Okay, I think this is ready to land! :tada:

shaunandrews commented:
I'm not sure what's going on, but my local install now shows a broken image instead of the WordPress logo:

<img width="162" alt="image" src="https://user-images.githubusercontent.com/191598/87713330-027d7200-c778-11ea-8f3d-4f566b0ee57c.png">

I don't think its related to this PR (I think my local install is busted in some other way), but it makes me think there should be a better fallback if the image doesn't exist, or if its been deleted from the library.

jeyip commented:
Thanks for letting us know Shaun.

The [code](https://github.com/WordPress/gutenberg/pull/22952/files#diff-5e54b8ed48d99cb46a591a186b9ad138R43), as it exists today, will use the WordPress Logo as the default icon to render if a user hasn't uploaded a custom one. I didn't realize that the default WordPress logo could be removed. Maybe the fallback should actually be to render nothing at all? Let me know what you think @shaunandrews

shaunandrews commented:
> I didn't realize that the default WordPress logo could be removed 

Sorry, I think there's a misunderstanding. I had a Site Icon setup previously, but at some point had deleted the image from my library. When this PR was merged, I noticed that the editor tried displaying my Site Icon, but the file no longer exists — hence the broken image in my previous comment.

youknowriad commented on lib/compat.php:
```diff
@@ -568,3 +568,20 @@ function gutenberg_output_html_nav_menu_item( $item_output, $item, $depth, $args
+ * @return array Modified path list to preload.
+ */
+function gutenberg_preload_edit_post( $preload_paths ) {
+	$additional_paths = array( '/?context=edit' );
```
This ticket has been rejected on Core, should we remove that code from here?

jeyip replied:
Thanks for the heads up Riad 👍 This is on my radar -- will respond when I get the chance tomorrow.

jeyip replied:
It's been quite a while since I've looked at this code, but yeah removing it makes sense to me, especially if, according to Timothy, the index endpoint doesn't even support `context` query params.

I'll spin up a draft to remove this code soon.

creativecoder replied:
Here's an attempt at making the entity request urls more specific to each entity, and prevents the need for preloading `/?context=edit`: https://github.com/WordPress/gutenberg/pull/30482
